go/engine/device_wrap.go

```diff
@@ -85,36 +85,34 @@ func (e *DeviceWrap) Run(m libkb.MetaContext) error {
 		PerUserKeyring: e.args.PerUserKeyring,
 	}
 	kgEng := NewDeviceKeygen(m.G(), kgArgs)
-	if err := RunEngine2(m, kgEng); err != nil {
+	if err = RunEngine2(m, kgEng); err != nil {
 		return err
 	}
 
 	pargs := &DeviceKeygenPushArgs{
 		Signer:    e.args.Signer,
 		EldestKID: e.args.EldestKID,
 	}
-	if err := kgEng.Push(m, pargs); err != nil {
+	if err = kgEng.Push(m, pargs); err != nil {
 		return err
 	}
 
 	e.signingKey = kgEng.SigningKey()
 	e.encryptionKey = kgEng.EncryptionKey()
 	// TODO get the per-user-key and save it if it was generated
 
-	if lctx := m.LoginContext(); lctx != nil {
-
-		// Set the device id so that SetCachedSecretKey picks it up.
-		// Signup does this too, but by then it's too late.
-		if err := lctx.LocalSession().SetDeviceProvisioned(deviceID); err != nil {
-			// Not fatal. Because, um, it was working ok before.
-			m.CWarningf("error saving session file: %s", err)
-		}
+	err = m.ActiveDevice().Set(m, e.args.Me.GetUID(), deviceID, e.signingKey, e.encryptionKey, e.args.DeviceName)
+	if err != nil {
+		return err
+	}
 
-		device := kgEng.device()
+	// Sync down secrets for future offline login attempts to work.
+	// This will largely just download what we just uploaded, but it's
+	// easy to do this way.
+	w := m.ActiveDevice().SyncSecrets(m)
```

Comment:
This looks a bit odd to me using `m` to get `ActiveDevice` then passing same `m` into SyncSecrets...I guess it's fine...